Flatten the module manifest structure (#188)
PowerShell/PowerShell-RFC: 1-Draft/RFCNNNN-Simplify-Module-Manifests.md

---
RFC: RFCnnnn
Author: Kirk Munro
Status: Draft
SupercededBy: 
Version: 1.0
Area: Module Manifests
Comments Due: July 15, 2019
Plan to implement: Yes
---

> # Title

Reviewer: ```suggestion
# Introduction
```

Author: Whoops, that's what I get for filling out an RFC late at night when I'm tired. Fixed. Thanks.

When PowerShell module support first came out in PowerShell 2.0, module manifests had a specific list of keys that could be used to define the manifest. Additional keys were added in PowerShell 3.0, but that hurt the end user experience because at the time, if you loaded a module with a manifest containing keys that were not recognized/supported in your version of PowerShell, PowerShell would return an unhelpful error about the module keys being invalid instead of checking for the minimum required version of PowerShell and returning an error indicating that the module only supports that version or later.

> Beyond PowerShell 3.0, keys started being added to a PSData subsection of the PrivateData area of modules. You can see these keys in use by the PowerShell Gallery. That approach has its own challenges, because now keys are defined in a hierarchy, which adds complexity and should not be necessary.

Reviewer: Please add examples (links to examples) - this helps reviewers.

Author: Will submit an update shortly.

Fast forward to today, and things have changed enough that we can reconsider how we set up a module manifest. `Import-Module` has been checking the required version of PowerShell before validating key names for several years now (at least since PowerShell 5.1).  Earlier versions of Windows PowerShell are no longer on mainstream support.

> With all of this in mind, I propose that we flatten the module manifest structure, reading values for keys at the root first, and then if they are not set at the root, looking in the PSData section under PrivateData for backwards compatibility support. Future keys added to the manifest as part of PowerShell should all be added to the root, leaving `PrivateData` for 3rd party or user information in a manifest, like it was intended.

Reviewer: ```suggestion
With all of this in mind, a proposal is to flatten the module manifest structure, reading values for keys at the root first, and then if they are not set at the root, looking in the PSData section under PrivateData for backwards compatibility support. Future keys added to the manifest as part of PowerShell should all be added to the root, leaving `PrivateData` for 3rd party or user information in a manifest, like it was intended.
```

Author: Resolved with different wording.

## Motivation

    As a scripter,
    I can generate manifests using `New-ModuleManifest` with all keys stored at the root of the hashtable,
    so that I can work with module manifests more easily.

## User Experience

> Here's an example showing how `New-ModuleManifest` would work after this change:

Reviewer: For greater clarity we could add commented `PSData` section with comment that these keys is moved out.
We could implement this to inform users about the change.

Author: Added a note to alternate proposals about this.

# Variables to export from this module
VariablesToExport = '*'

# Aliases to export from this module, for best performance, do not use wildcards and do not delete the entry, use an empty array if there are no aliases to export.
AliasesToExport = @()

# DSC resources to export from this module
# DscResourcesToExport = @()

# List of all modules packaged with this module
# ModuleList = @()

# List of all files packaged with this module
# FileList = @()

# Private data to pass to the module specified in RootModule/ModuleToProcess.
PrivateData = @{}

# HelpInfo URI of this module
# HelpInfoURI = ''

# Default prefix for commands exported from this module. Override the default prefix using Import-Module -Prefix.
# DefaultCommandPrefix = ''

# Tags applied to this module. These help with module discovery in online galleries.
# Tags = @()

# A URL to the license for this module.
# LicenseUri = ''

# A URL to the main website for this project.
# ProjectUri = ''

# A URL to an icon representing this module.
# IconUri = ''

# ReleaseNotes of this module
# ReleaseNotes = ''

}

```

## Specification

The changes required for this RFC are relatively straightforward, and as follows:

> 1. `New-ModuleManifest` would generate a template like what is shown above, omitting any details about the `PSData` section.

Reviewer: Should we explicitly say that we deprecate the `PSData` section?

Author: I think I have this covered in the updated alternate proposals details.

1. Module import logic would read the manifest and pull the values for `Tags`, `LicenseUri`, `ProjectUri`, `IconUri` and `ReleaseNotes` from the top-level keys. If top-level keys were not defined with these values, it would look for values in a `PSData` section for backward compatibility.
1. If values are defined in both locations, an error would occur informing the module author that they need to remove one of the two keys.
1. Update `Test-ModuleManifest` to validate the keys in the new locations.
1. Update `Update-ModuleManifest` to set the keys in the existing location if it is in use, or in the new location otherwise.

## Alternate Proposals and Considerations

### Move `Update-ModuleManifest` from PowerShellGet to Microsoft.PowerShell.Core

> This is a question for consideration: why is `Update-ModuleManifest` part of PowerShellGet and not in the Microsoft.PowerShell.Core module? Shouldn't it be co-located with the other `*-ModuleManifest` cmdlets? If so, it could be moved as part of this RFC.

Reviewer: It would be nice. Perhaps PowerShellGet 3.0 RFC is better place to discuss the propose.

Author: Updated this text, because I realized having it in PowerShellGet gives downlevel support. I still think they should be co-located, though, because there is too much of an update risk having them in separate modules.

On the down side, it requires the management of additional files when working on manifests, which isn't necessarily simplifying manifest management.

### Add a `-BackwardCompatible` parameter to `New-ModuleManifest`

@iSazonov proposed we consider adding a `-BackwardCompatible` switch parameter to `New-ModuleManifest`. When that parameter is used, the manifest would be generated in a backward compatible format with new values being stored in `PrivateData`. Otherwise, it would generate the manifest in the latest format with new values stored at the root.

> Based on how parameters have been added in recent versions, that would work for now, but going forward how would this work? For example, if one or more new keys are added to PowerShell 7, and then others in 7.x or 8, how would you generate a manifest to target the appropriate downlevel version? For that reason, I don't believe this is a good option to move forward with.

Reviewer: My thoughts was about compatibility versions < 5.1, `-BackwardCompatibleVer3`.
Recent version and newest ones (as I understand) will resolve this 
by themselves.

Author: I see, sorry that I misunderstood. So assuming that future versions of PowerShell (7+) allow for additional keys to be read from the root folder and perhaps not used without generating an error, that would work. I'll update that section.
